File: khalti-android/src/main/java/khalti/checkOut/Wallet/Wallet.java
package khalti.checkOut.Wallet;

import android.Manifest;
import android.app.Dialog;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.CardView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.jakewharton.rxbinding.view.RxView;
import com.jakewharton.rxbinding.widget.RxTextView;

import java.util.HashMap;

import khalti.R;
import khalti.SmsListener;
import khalti.carbonX.widget.Button;
import khalti.carbonX.widget.FrameLayout;
import khalti.utils.AppPermissionUtil;
import khalti.utils.EmptyUtil;
import khalti.utils.ExpandableLayout;
import khalti.utils.NetworkUtil;
import khalti.utils.NumberUtil;
import khalti.utils.ResourceUtil;
import khalti.utils.Store;
import khalti.utils.UserInterfaceUtil;
import rx.Observable;

public class Wallet extends Fragment implements khalti.checkOut.Wallet.WalletContract.View {

    private EditText etMobile, etCode, etPIN;
    private khalti.carbonX.widget.TextInputLayout tilMobile, tilCode, tilPIN;
    private ExpandableLayout elConfirmation;
    private Button btnPay;
    private Dialog progressDialog;
    private LinearLayout llConfirmation, llPIN, llCode, llKhaltiBranding;
    private ImageView ivKhalti;
    private CardView cvPinMessage;
    private AppCompatTextView tvPinMessage;

    private FragmentActivity fragmentActivity;
    private WalletContract.Presenter presenter;
    private SmsListener smsListener;

    private boolean isRegistered = false;
    private int height = 0;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View mainView = inflater.inflate(R.layout.wallet_form, container, false);
        fragmentActivity = getActivity();
        presenter = new WalletPresenter(this);

        etMobile = mainView.findViewById(R.id.etMobile);
        etCode = mainView.findViewById(R.id.etCode);
        etPIN = mainView.findViewById(R.id.etPIN);
        tilMobile = mainView.findViewById(R.id.tilMobile);
        tilCode = mainView.findViewById(R.id.tilCode);
        tilPIN = mainView.findViewById(R.id.tilPIN);
        btnPay = mainView.findViewById(R.id.btnPay);
        elConfirmation = mainView.findViewById(R.id.elConfirmation);
        llConfirmation = mainView.findViewById(R.id.llConfirmation);
        llPIN = mainView.findViewById(R.id.llPIN);
        llCode = mainView.findViewById(R.id.llCode);
        ivKhalti = mainView.findViewById(R.id.ivKhalti);
        llKhaltiBranding = mainView.findViewById(R.id.llKhaltiBranding);
        cvPinMessage = mainView.findViewById(R.id.cvPinMessage);
        tvPinMessage = mainView.findViewById(R.id.tvPinMessage);

        presenter.onCreate();

        return mainView;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.onDestroy();
    }

    @Override
    public HashMap<String, Observable<CharSequence>> setEditTextListener() {
        return new HashMap<String, Observable<CharSequence>>() {{
            put("mobile", RxTextView.textChanges(etMobile));
            put("code", RxTextView.textChanges(etCode));
            put("pin", RxTextView.textChanges(etPIN));
        }};
    }

    @Override
    public void toggleProgressDialog(String action, boolean show) {
        FrameLayout flCircularProgress = (FrameLayout) fragmentActivity.getLayoutInflater().inflate(R.layout.component_circular_progress, null);
        String message = "";
        switch (action) {
            case "init":
                message = ResourceUtil.getString(fragmentActivity, R.string.init_payment);
                break;
            case "confirm":
                message = ResourceUtil.getString(fragmentActivity, R.string.confirming_payment);
                break;
        }

        if (show) {
            progressDialog = UserInterfaceUtil.runProgressDialog(fragmentActivity, message, ResourceUtil.getString(fragmentActivity, R.string.please_wait), flCircularProgress, () -> {
//                presenter.unSubscribe();
            });
        } else {
            if (EmptyUtil.isNotNull(progressDialog)) {
                progressDialog.dismiss();
            }
        }
    }

    @Override
    public void setEditTextError(String view, String error) {
        boolean isError = EmptyUtil.isNotNull(error);
        switch (view) {
            case "mobile":
                tilMobile.setErrorEnabled(isError);
                tilMobile.setError(error);
                break;
            case "code":
                llCode.measure(0, 0);
                int beforeA = llCode.getMeasuredHeight();

                tilCode.setErrorEnabled(isError);
                tilCode.setError(error);

                llCode.measure(0, 0);
                int afterA = llCode.getMeasuredHeight();

                height = Math.abs(height + (afterA - beforeA));
                break;
            case "pin":
                llPIN.measure(0, 0);
                int beforeB = llPIN.getMeasuredHeight();

                tilPIN.setErrorEnabled(isError);
                tilPIN.setError(error);

                llPIN.measure(0, 0);
                int afterB = llPIN.getMeasuredHeight();

                height = Math.abs(height + (afterB - beforeB));
                break;
        }
    }

    @Override
    public void setButtonText(String text) {
        btnPay.setText(text);
    }

    @Override
    public Observable<Void> setButtonClickListener() {
        return RxView.clicks(btnPay);
    }

    @Override
    public Observable<Void> setImageClickListener() {
        return RxView.clicks(ivKhalti);
    }

    @Override
    public void setConfirmationCode(String code) {
        etCode.setText(code);
        etPIN.requestFocus();
    }

    @Override
    public void setConfirmationLayoutHeight(String view) {
        TextInputLayout til;
        LinearLayout ll;
        if (view.equals("code")) {
            til = tilCode;
            ll = llCode;
        } else {
            til = tilPIN;
            ll = llPIN;
        }
        if (EmptyUtil.isNotNull(til.getError())) {
            ll.measure(0, 0);
            int beforeA = ll.getMeasuredHeight();

            til.setErrorEnabled(false);

            ll.measure(0, 0);
            int afterA = ll.getMeasuredHeight();

            height = Math.abs(height + (beforeA - afterA));

            ExpandableLayout.LayoutParams layoutParams = (ExpandableLayout.LayoutParams) llConfirmation.getLayoutParams();
            layoutParams.height = llConfirmation.getHeight() - height;
            llConfirmation.setLayoutParams(layoutParams);

            height = 0;
        }
    }

    @Override
    public void showNetworkError() {
        UserInterfaceUtil.showSnackBar(fragmentActivity, ((khalti.checkOut.CheckOutActivity) this.fragmentActivity).cdlMain, ResourceUtil.getString(fragmentActivity, R.string.network_error_body),
                false, "", 0, 0, null);
    }

    @Override
    public void showMessageDialog(String title, String message) {
        UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, true, true, ResourceUtil.getString(fragmentActivity, R.string.got_it), null,
                new UserInterfaceUtil.DialogAction() {
                    @Override
                    public void onPositiveAction(Dialog dialog) {
                        dialog.dismiss();
                    }

                    @Override
                    public void onNegativeAction(Dialog dialog) {

                    }
                });
    }

    @Override
    public void showPINDialog(String title, String message) {
        UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, true, true, ResourceUtil.getString(fragmentActivity, R.string.ok),
                ResourceUtil.getString(fragmentActivity, R.string.cancel), new UserInterfaceUtil.DialogAction() {
                    @Override
                    public void onPositiveAction(Dialog dialog) {
                        dialog.dismiss();
                        presenter.openKhaltiSettings();
                    }

                    @Override
                    public void onNegativeAction(Dialog dialog) {
                        dialog.dismiss();
                    }
                });
    }

    @Override
    public void showPINInBrowserDialog(String title, String message) {
        UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, true, true, ResourceUtil.getString(fragmentActivity, R.string.ok),
                ResourceUtil.getString(fragmentActivity, R.string.cancel), new UserInterfaceUtil.DialogAction() {
                    @Override
                    public void onPositiveAction(Dialog dialog) {
                        dialog.dismiss();
                        presenter.openLinkInBrowser();
                    }

                    @Override
                    public void onNegativeAction(Dialog dialog) {
                        dialog.dismiss();
                    }
                });
    }

    @Override
    public void openKhaltiSettings() {
        Intent i;
        PackageManager manager = fragmentActivity.getPackageManager();
        try {
            i = manager.getLaunchIntentForPackage("com.khalti.red");
            i = EmptyUtil.isNotNull(i) ? i : manager.getLaunchIntentForPackage("com.khalti");
            if (EmptyUtil.isNull(i)) {
                throw new PackageManager.NameNotFoundException();
            }
            i.addCategory(Intent.CATEGORY_LAUNCHER);

            HashMap<String, String> map = new HashMap<>();
            map.put("pin_settings", "pin_settings");
            i.putExtra("map", map);

            startActivity(i);
        } catch (PackageManager.NameNotFoundException e) {
            presenter.showPINInBrowserDialog("Error", ResourceUtil.getString(fragmentActivity, R.string.khalti_not_found) + "\n\n" +
                    ResourceUtil.getString(fragmentActivity, R.string.set_pin_in_browser));
        }
    }

    @Override
    public void openLinkInBrowser(String link) {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(link));
        startActivity(browserIntent);
    }

    @Override
    public void closeWidget() {
        fragmentActivity.finish();
    }

    @Override
    public void updateConfirmationHeight() {
        ExpandableLayout.LayoutParams layoutParams = (ExpandableLayout.LayoutParams) llConfirmation.getLayoutParams();
        layoutParams.height = llConfirmation.getHeight() + height;
        llConfirmation.setLayoutParams(layoutParams);

        height = 0;
    }

    @Override
    public String getMessage(String action) {
        switch (action) {
            case "pin_not_set":
                return ResourceUtil.getString(fragmentActivity, R.string.pin_not_set);
            case "pin_not_set_continue":
                return ResourceUtil.getString(fragmentActivity, R.string.pin_not_set_continue);
            default:
                return "";
        }
    }

    @Override
    public boolean hasSmsReceiptPermission() {
        return AppPermissionUtil.checkAndroidPermission(fragmentActivity, Manifest.permission.RECEIVE_SMS);
    }

    @Override
    public void askSmsReceiptPermission() {
        AppPermissionUtil.askPermission(fragmentActivity, Manifest.permission.RECEIVE_SMS, "Please allow permission to receive SMS", () -> presenter.onSmsReceiptPermitted());
    }

    @Override
    public boolean hasNetwork() {
        return NetworkUtil.isNetworkAvailable(fragmentActivity);
    }

    @Override
    public String getPayButtonText() {
        return btnPay.getText() + "";
    }

    @Override
    public HashMap<String, String> getFormData() {
        return new HashMap<String, String>() {{
            put("mobile", etMobile.getText() + "");
            put("code", etCode.getText() + "");
            put("pin", etPIN.getText() + "");
        }};
    }

    @Override
    public void showSlogan() {
        Toast.makeText(fragmentActivity, ResourceUtil.getString(fragmentActivity, R.string.slogan), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showBranding() {
        llKhaltiBranding.setVisibility(View.VISIBLE);
    }

    @Override
    public void toggleConfirmationLayout(boolean show) {
        String buttonText = show ? ResourceUtil.getString(fragmentActivity, R.string.confirm_payment) : "Pay Rs " + NumberUtil.convertToRupees(Store.getConfig().getAmount());
        btnPay.setText(buttonText);
        etCode.setText("");
        etPIN.setText("");
        elConfirmation.toggleExpansion();
    }

    @Override
    public void toggleSmsListener(boolean listen) {
        if (listen) {
            IntentFilter intentFilter = new IntentFilter("android.provider.Telephony.SMS_RECEIVED");
            smsListener = new SmsListener();
            fragmentActivity.registerReceiver(smsListener, intentFilter);
            isRegistered = true;
        } else {
            if (EmptyUtil.isNotNull(smsListener) && isRegistered) {
                fragmentActivity.unregisterReceiver(smsListener);
                isRegistered = false;
            }
        }
    }

    @Override
    public void togglePinMessage(boolean show) {
        cvPinMessage.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override
    public void setPinMessage(String message) {
        String s = message + " " + ResourceUtil.getString(fragmentActivity, R.string.sms_info);
        tvPinMessage.setText(s);
    }

    @Override
    public void setMobile(String mobile) {
        etMobile.setText(mobile);
        etMobile.setSelection(mobile.length());
    }

    @Override
    public void setPresenter(WalletContract.Presenter presenter) {
        this.presenter = presenter;
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/wallet/Wallet.kt
package com.khalti.checkOut.wallet

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.Toast
import androidx.appcompat.app.AppCompatDialog
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputLayout
import com.khalti.R
import com.khalti.signal.Signal
import com.khalti.utils.*
import kotlinx.android.synthetic.main.wallet_form.*
import java.util.*
import kotlin.math.abs

class Wallet : Fragment(), WalletContract.View {

    private var progressDialog: AppCompatDialog? = null
    private lateinit var fragmentActivity: FragmentActivity
    private lateinit var presenter: WalletContract.Presenter

    private var height: Int = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val mainView = inflater.inflate(R.layout.wallet_form, container, false)
        fragmentActivity = activity!!
        presenter = WalletPresenter(this)

        presenter.onCreate()

        return mainView
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.onDestroy()
    }

    override val payButtonText: String get() = ViewUtil.getText(btnPay)

    override val formData: Map<String, String>
        get() = object : HashMap<String, String>() {
            init {
                put("mobile", ViewUtil.getText(etMobile))
                put("code", ViewUtil.getText(etCode))
                put("pin", ViewUtil.getText(etPIN))
            }
        }

    override fun toggleProgressDialog(action: String, show: Boolean): Signal<Boolean> {
        var signal = Signal<Boolean>()
        var message = ""
        when (action) {
            "init" -> message = ResourceUtil.getString(fragmentActivity, R.string.init_payment)
            "confirm" -> message = ResourceUtil.getString(fragmentActivity, R.string.confirming_payment)
        }

        if (show) {
            val pair = UserInterfaceUtil.runProgressDialog(fragmentActivity, message, ResourceUtil.getString(fragmentActivity, R.string.please_wait))
            progressDialog = pair.first
            signal = pair.second
        } else {
            progressDialog?.dismiss()
        }
        return signal
    }

    override fun toggleConfirmationLayout(show: Boolean) {
        val buttonText = if (show) ResourceUtil.getString(fragmentActivity, R.string.confirm_payment) else "Pay Rs ${NumberUtil.convertToRupees(Store.getConfig().amount)}"
        btnPay?.text = buttonText
        etCode?.setText("")
        etPIN?.setText("")
        elConfirmation.toggleExpansion()
    }

    override fun togglePinMessage(show: Boolean) {
        cvPinMessage?.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun setPinMessage(message: String) {
        tvPinMessage?.text = message + " " + ResourceUtil.getString(fragmentActivity, R.string.sms_info)
    }

    override fun setMobile(mobile: String) {
        etMobile?.setText(mobile)
        etMobile?.setSelection(mobile.length)
    }

    override fun setEditTextError(view: String, error: String?) {
        val isError = EmptyUtil.isNotNull(error)
        when (view) {
            "mobile" -> {
                tilMobile?.isErrorEnabled = isError
                tilMobile?.error = error
            }
            "code" -> {
                llCode.measure(0, 0)
                val beforeA = llCode.measuredHeight

                tilCode.isErrorEnabled = isError
                tilCode.error = error

                llCode.measure(0, 0)
                val afterA = llCode.measuredHeight

                height = abs(height + (afterA - beforeA))
            }
            "pin" -> {
                llPIN.measure(0, 0)
                val beforeB = llPIN.measuredHeight

                tilPIN.isErrorEnabled = isError
                tilPIN.error = error

                llPIN.measure(0, 0)
                val afterB = llPIN.measuredHeight

                height = abs(height + (afterB - beforeB))
            }
        }
    }

    override fun setButtonText(text: String) {
        btnPay?.text = text
    }

    override fun setConfirmationLayoutHeight(view: String) {
        val til: TextInputLayout
        val ll: LinearLayout
        if (view == "code") {
            til = tilCode
            ll = llCode
        } else {
            til = tilPIN
            ll = llPIN
        }
        if (EmptyUtil.isNotNull(til?.error)) {
            ll?.measure(0, 0)
            val beforeA = ll?.measuredHeight

            til?.setErrorEnabled(false)

            ll?.measure(0, 0)
            val afterA = ll?.measuredHeight

            if (EmptyUtil.isNotNull(beforeA) && EmptyUtil.isNotNull(afterA)) {
                height = abs(height + (beforeA!! - afterA!!))

                val layoutParams = llConfirmation.layoutParams as ExpandableLayout.LayoutParams
                layoutParams.height = llConfirmation.height - height
                llConfirmation.layoutParams = layoutParams

                height = 0
            }
        }
    }

    override fun setEditTextListener(): Map<String, Signal<CharSequence>> = object : HashMap<String, Signal<CharSequence>>() {
        init {
            put("mobile", ViewUtil.setTextChangeListener(etMobile))
            put("code", ViewUtil.setTextChangeListener(etCode))
            put("pin", ViewUtil.setTextChangeListener(etPIN))
        }
    }

    override fun setClickListener(): Map<String, Signal<Any>> = object : HashMap<String, Signal<Any>>() {
        init {
            put("pay", ViewUtil.setClickListener(btnPay))
            put("khalti", ViewUtil.setClickListener(ivKhalti))
        }
    }

    override fun showNetworkError() {
        UserInterfaceUtil.showSnackBar(fragmentActivity, (this.fragmentActivity as com.khalti.checkOut.CheckOutActivity).cdlMain, ResourceUtil.getString(fragmentActivity, R.string.network_error_body), null, Snackbar.LENGTH_LONG)
    }

    override fun showMessageDialog(title: String, message: String) {
        UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, autoDismiss = true, cancelable = true, positiveText = ResourceUtil.getString(fragmentActivity, R.string.got_it), negativeText = null)
    }

    override fun showPINDialog(title: String, message: String): Signal<Boolean> =
            UserInterfaceUtil.showInfoDialog(fragmentActivity, title, message, autoDismiss = true, cancelable = true, positiveText = ResourceUtil.getString(fragmentActivity, R.string.ok), negativeText = ResourceUtil.getString(fragmentActivity, R.string.cancel))

    override fun showSlogan() {
        Toast.makeText(fragmentActivity, ResourceUtil.getString(fragmentActivity, R.string.slogan), Toast.LENGTH_SHORT).show()
    }

    override fun showBranding() {
        llKhaltiBranding.visibility = View.VISIBLE
    }

    override fun openKhaltiSettings() {
        var i: Intent?
        val manager = fragmentActivity.packageManager
        try {
            i = manager.getLaunchIntentForPackage("com.khalti.red")
            i = if (EmptyUtil.isNotNull(i)) i else manager.getLaunchIntentForPackage("com.khalti")
            if (EmptyUtil.isNull(i)) {
                throw PackageManager.NameNotFoundException()
            }
            i!!.addCategory(Intent.CATEGORY_LAUNCHER)

            val map = HashMap<String, String>()
            map["pin_settings"] = "pin_settings"
            i.putExtra("map", map)

            startActivity(i)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()

            /*presenter.showPINInBrowserDialog("Error", ResourceUtil.getString(fragmentActivity, R.string.khalti_not_found) + "\n\n" +
                    ResourceUtil.getString(fragmentActivity, R.string.set_pin_in_browser))*/
        }

    }

    override fun openLinkInBrowser(link: String) {
        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(link))
        startActivity(browserIntent)
    }

    override fun closeWidget() {
        fragmentActivity.finish()
    }

    override fun updateConfirmationHeight() {
        val layoutParams = llConfirmation.layoutParams as ExpandableLayout.LayoutParams
        layoutParams.height = llConfirmation.height + height
        llConfirmation.layoutParams = layoutParams

        height = 0
    }

    override fun getMessage(action: String): String {
        return when (action) {
            "pin_not_set" -> ResourceUtil.getString(fragmentActivity, R.string.pin_not_set)
            "pin_not_set_continue" -> ResourceUtil.getString(fragmentActivity, R.string.pin_not_set_continue)
            "khalti_not_found" -> ResourceUtil.getString(fragmentActivity, R.string.khalti_not_found)
            "set_pin_in_browser" -> ResourceUtil.getString(fragmentActivity, R.string.set_pin_in_browser)
            else -> ""
        }
    }

    override fun hasNetwork(): Boolean {
        return NetworkUtil.isNetworkAvailable(fragmentActivity)
    }

    override fun doesPackageExist(): Boolean {
        val manager = fragmentActivity.packageManager
        return try {
            var i = manager.getLaunchIntentForPackage("com.khalti.red")
            if (EmptyUtil.isNull(i)) {
                i = manager.getLaunchIntentForPackage("com.khalti")
            }

            EmptyUtil.isNotNull(i)

        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            false
        }
    }

    override fun setPresenter(presenter: WalletContract.Presenter) {
        this.presenter = presenter
    }
}